File: clients/rs/src/features/config_env_variables.rs
use polywrap_client::{
    builder::{PolywrapClientConfig, PolywrapClientConfigBuilder},
    client::PolywrapClient,
    core::{invoker::Invoker, uri::Uri},
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::error::Error;

#[derive(Serialize, Deserialize)]
struct Env {
  str: String,
  num: u8
}

#[derive(Deserialize)]
struct InputObj {
    uri: String,
    env: Env,
}

pub fn run_test_case(input: &Value) -> Result<(), Box<dyn Error>> {
    let input_obj: InputObj = serde_json::from_value(input.clone())?;
    let uri: Uri = input_obj.uri.try_into()?;

    println!("Adding Env to ClientConfig");

    let mut config = PolywrapClientConfig::new();
    config.add_env(
        uri.clone(),
        polywrap_client::msgpack::to_vec(&input_obj.env)?,
    );

    let client: PolywrapClient = PolywrapClient::new(config.into());

    println!("Fetching Env");

    let result = client
        .get_env_by_uri(&uri);

    if let Some(result) = result {
        let result = polywrap_client::msgpack::from_slice::<Env>(&result)?;

        // TODO: section altered because of key-ordering

        let value = result.str;
        println!("env.str = {}", value);

        let value = result.num;
        println!("env.num = {}", value);

        println!("Success!")
    }

    Ok(())
}

File: clients/swift/Sources/Readiness/Features/ConfigEnvVariables.swift
import PolywrapClient
import Foundation

public struct CustomEnv: Codable {
    var str: String
    var num: Int
    
    public init(str: String, num: Int) {
        self.str = str
        self.num = num
    }
}

struct ConfigEnvVariablesTest: Feature {
    func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any] else {
           throw InputError.expectedObject
        }

        let uri = try Input.expectUri(inputObj["uri"])

        guard let envDict = inputObj["env"] as? [AnyHashable: Any],
             let str = envDict["str"] as? String,
             let num = envDict["num"] as? Int else {
           throw InputError.expectedObject
        }

        let env = Env(str: str, num: num)

        print("Adding Env to ClientConfig")

        let builder = BuilderConfig()
        try builder.addEnv(uri, env)
        let client = builder.build()

        print("Fetching Env")

        guard let result: CustomEnv = client.getEnvByUri(uri) else {
            let error = NSError(domain: "", code: 1, userInfo: [NSLocalizedDescriptionKey : "No environment variables found for \(uri)"])
            throw error
        }

        print("env.str = \(result.str)")
        print("env.num = \(result.num)")
        print("Success!")
    }
}
